core/divesitehelpers.cpp

```diff
@@ -48,119 +46,115 @@ void reverseGeoLookup(degrees_t latitude, degrees_t longitude, taxonomy_data *ta
+			int ocean;
+			if ((ocean = taxonomy_index_for_category(taxonomy, TC_OCEAN)) > 0) {
+				taxonomy->category[0] = taxonomy->category[ocean];
+				taxonomy->nr = 1;
 			} else {
-				// clear out the data (except for the ocean data)
-				int ocean;
-				if ((ocean = taxonomy_index_for_category(taxonomy, TC_OCEAN)) > 0) {
-					taxonomy->category[0] = taxonomy->category[ocean];
-					taxonomy->nr = 1;
-				} else {
-					// ocean is -1 if there is no such entry, and we didn't copy above
-					// if ocean is 0, so the following gets us the correct count
-					taxonomy->nr = ocean + 1;
-				}
-			}
-			// get all the data - OCEAN is special, so start at COUNTRY
-			for (int j = TC_COUNTRY; j < TC_NR_CATEGORIES; j++) {
-				if (firstData[taxonomy_api_names[j]].isValid()) {
-					taxonomy->category[ri].category = j;
-					taxonomy->category[ri].origin = taxonomy_origin::GEOCODED;
-					free((void *)taxonomy->category[ri].value);
-					taxonomy->category[ri].value = copy_qstring(firstData[taxonomy_api_names[j]].toString());
-					ri++;
-				}
+				// ocean is -1 if there is no such entry, and we didn't copy above
+				// if ocean is 0, so the following gets us the correct count
+				taxonomy->nr = ocean + 1;
 			}
-			taxonomy->nr = ri;
-			l3 = taxonomy_index_for_category(taxonomy, TC_ADMIN_L3);
-			lt = taxonomy_index_for_category(taxonomy, TC_LOCALNAME);
-			if (l3 == -1 && lt != -1) {
-				// basically this means we did get a local name (what we call town), but just like most places
-				// we didn't get an adminName_3 - which in some regions is the actual city that town belongs to,
-				// then we copy the town into the city
-				taxonomy->category[ri].value = copy_string(taxonomy->category[lt].value);
-				taxonomy->category[ri].origin = taxonomy_origin::GEOCOPIED;
-				taxonomy->category[ri].category = TC_ADMIN_L3;
-				taxonomy->nr++;
+		}
+		// get all the data - OCEAN is special, so start at COUNTRY
+		for (int j = TC_COUNTRY; j < TC_NR_CATEGORIES; j++) {
+			if (firstData[taxonomy_api_names[j]].isValid()) {
+				taxonomy->category[ri].category = j;
+				taxonomy->category[ri].origin = taxonomy_origin::GEOCODED;
+				free(const_cast<char*>(taxonomy->category[ri].value));
+				taxonomy->category[ri].value = copy_qstring(firstData[taxonomy_api_names[j]].toString());
+				ri++;
 			}
-		} else {
-			report_error("geonames.org did not provide reverse lookup information");
-			qDebug() << "no reverse geo lookup; geonames returned\n" << fullReply;
+		}
+		taxonomy->nr = ri;
+		l3 = taxonomy_index_for_category(taxonomy, TC_ADMIN_L3);
+		lt = taxonomy_index_for_category(taxonomy, TC_LOCALNAME);
+		if (l3 == -1 && lt != -1) {
+			// basically this means we did get a local name (what we call town), but just like most places
+			// we didn't get an adminName_3 - which in some regions is the actual city that town belongs to,
+			// then we copy the town into the city
+			taxonomy->category[ri].value = copy_string(taxonomy->category[lt].value);
```

Comment:
In the code above, the old string is freed but not here. Is this on purpose?

Reply:
> Overall might be better to entirely discard the old taxonomy and start from a new one. Or to actually clear out the data where it says it does. And move the OCEAN to the front to stop it being such a special case.

Or, since this is called from C++ only, simply return a `taxonomy_data` structure instead of passing it in. Thus it becomes clear who is responsible for object construction.

BTW: It just occurs to me that your last PR introduced a (very minor) resource leak:

```
	taxonomy_data taxonomy = { 0, 0 };
	reverseGeoLookup(location.lat, location.lon, &taxonomy);
	if (ds != diveSite)
		return; // <--- taxonomy data is now leaked!
```

C-style memory management is so pointlessly painful. :-/

What I did in the undo commands to make this less painful is to introduce "owning pointers", which are `unique_ptr`s with a deleter function. You could then declare the function as
```
OwningTaxonomyPtr reverseGeoLookup(degrees_t latitude, degrees_t longitude);
```
and use it like thus:
```
    OwningTaxonomyPtr taxonomy = reverseGeoLookup(...);
    if (!taxonomy || ds != diveSite)
        return;
    Command::editDiveSiteTaxonomy(ds, *taxonomy);
```

The trick is: when the pointer goes out of scope, the taxonomy is automatically freed! Nothing to worry about!

Note: I explicitly do NOT suggest that you implement such a scheme, it was just a FYI.

Reply:
> Or, since this is called from C++ only, simply return a taxonomy_data structure instead of passing it in. 

That doesn't (historically) appear to have been the case, hence the rather convoluted code in the original function.  These days it's only called from a single place and yes, it would reduce the scope of the function and make it very clear exactly what it does (returns the reverse geocode lookup results and doesn't try anything "clever" by merging data etc)

> BTW: It just occurs to me that your last PR introduced a (very minor) resource leak:

Indeed..  *sigh*  Although that was someone's suggested approach, not my original solution (I didn't know that reverseGeoLookup() allocated memory at that point).

>What I did in the undo commands to make this less painful is to introduce "owning pointers", which are unique_ptrs with a deleter function. You could then declare the function as

Ok, how does that work? Given that taxonomy_data.category is a malloc'd pointer.  I'm guessing the "OwningTaxonomyPtr" is a wrapper-class which takes care of it?  Can you give an example of where you implemented one of these?

Reply:
Ok, so I found the 'OwningPtr' implementations, but the only way to implement it as you suggest is if reverseGeoLookup() allocates not only 'taxonomy->category', but the 'taxonomy_data' structure as well. That makes for even messier handling inside the function.

I think I'll have to leave it as per my latest commit (pending review) unless we want to completely revamp everything and turn things into proper C++ classes. Or change "taxonomy_data.category" into being an array inside the struct instead of a pointer, and then just allocate "taxonomy_data".

Which, I think, is getting well beyond the scope for this (originally relatively simple) pull request.  I'm sorry, but my time is quite limited otherwise I'd love to help with more comprehensive work.

Reply:
As I said - I'm not expecting you to implement such a scheme and I am fine with the `free_taxonomy` call. I will go through it once more today, but it should be fine now.